src/number_theory/number_field/canonical_embedding.lean

```diff
@@ -0,0 +1,171 @@
+        (λ w, (⟨w x, map_nonneg w x⟩ : nnreal))).symm using 2,
+      ext w,
+      simp only [apply_at_real_infinite_place, coe_nnnorm, real.norm_eq_abs,
+        function.embedding.coe_subtype, subtype.coe_mk, is_real.abs_embedding_apply], },
+    { convert (finset.univ.sup_map (function.embedding.subtype (λ w : infinite_place K,
+        is_complex w)) (λ w, (⟨w x, map_nonneg w x⟩ : nnreal))).symm using 2,
+      ext w,
+      simp only [apply_at_complex_infinite_place, subtype.val_eq_coe, coe_nnnorm,
+        complex.norm_eq_abs, function.embedding.coe_subtype, subtype.coe_mk, abs_embedding], }},
+  { ext w,
+    simp only [w.is_real_or_is_complex, set.mem_set_of_eq, finset.mem_union, set.mem_to_finset,
+      finset.mem_univ], },
+end
+
+lemma norm_le_iff [number_field K] (x : K) (r : ℝ) :
+  ‖canonical_embedding K x‖ ≤ r ↔ ∀ w : infinite_place K, w x ≤ r :=
+begin
+  obtain hr | hr := lt_or_le r 0,
+  { obtain ⟨w⟩ := infinite_place.nonempty K,
+    exact iff_of_false (hr.trans_le $ norm_nonneg _).not_le
+      (λ h, hr.not_le $ (map_nonneg w _).trans $ h _) },
+  { lift r to nnreal using hr,
+    simp_rw [← coe_nnnorm, nnnorm_eq, nnreal.coe_le_coe, finset.sup_le_iff, finset.mem_univ,
+      forall_true_left, ←nnreal.coe_le_coe, subtype.coe_mk] }
+end
+
+variables (K)
+
+/-- The image of `𝓞 K` as a subring of `ℝ^r₁ × ℂ^r₂`. -/
+def integer_lattice : subring (space K) :=
+(ring_hom.range (algebra_map (𝓞 K) K)).map (canonical_embedding K)
+
+/-- The ring equiv between `𝓞 K` and the integer lattice. -/
+def equiv_integer_lattice [number_field K] :
+  𝓞 K ≃ₗ[ℤ] integer_lattice K :=
+linear_equiv.of_bijective
+  { to_fun := λ x, ⟨canonical_embedding K (algebra_map (𝓞 K) K x), algebra_map (𝓞 K) K x,
+      by simp only [subring.mem_carrier, ring_hom.mem_range, exists_apply_eq_apply], rfl⟩,
+    map_add' := λ x y, by simpa only [map_add],
+    map_smul' := λ c x, by simpa only [zsmul_eq_mul, map_mul, map_int_cast] }
+  begin
+    refine ⟨λ _ _ h, _,  λ ⟨_, _, ⟨a, rfl⟩, rfl⟩, ⟨a, rfl⟩⟩,
+    rw [linear_map.coe_mk, subtype.mk_eq_mk] at h,
+    exact is_fraction_ring.injective (𝓞 K) K (canonical_embedding_injective K h),
+  end
+
+lemma integer_lattice.inter_ball_finite [number_field K] (r : ℝ) :
+  ((integer_lattice K : set (space K)) ∩ (closed_ball 0 r)).finite :=
+begin
+  obtain hr | hr := lt_or_le r 0,
+  {  simp [closed_ball_eq_empty.2 hr] },
+  have heq :
+    ∀ x, canonical_embedding K x ∈ closed_ball (0 : (space K)) r ↔ ∀ φ : K →+* ℂ, ‖φ x‖ ≤ r,
+  { simp only [← place_apply, ← infinite_place.coe_mk, mem_closed_ball_zero_iff, norm_le_iff],
+    exact λ x, le_iff_le x r, },
+  convert (embeddings.finite_of_norm_le K ℂ r).image (canonical_embedding K),
+  ext, split,
+  { rintro ⟨⟨_, ⟨x, rfl⟩, rfl⟩, hx2⟩,
+    exact ⟨x, ⟨set_like.coe_mem x, (heq x).mp hx2⟩, rfl⟩, },
+  { rintro ⟨x, ⟨hx1, hx2⟩, rfl⟩,
+    exact ⟨⟨x, ⟨⟨x, hx1⟩, rfl⟩, rfl⟩, (heq x).mpr hx2⟩, }
+end
+
+instance [number_field K] : countable (integer_lattice K) :=
```

Comment:
Sorry I don't have lean open so maybe I'm misunderstood something about the setup but:
Surely the integer lattice being countable follows from the fact the number field itself is countable? So it seems this result doesn't really belong here or require the canonical embedding.

Reply:
Yes, this is true and one can prove that the `unit_lattice` is countable that way. However, the important result here is `integer_lattice.inter_ball_finite` just above that is needed to prove that it is a full lattice. The fact that this proves that it is also countable is a by product and we will need this result later on to apply Minkowski.